ext/cuda/remapping_distributed.jl

```diff
@@ -60,41 +61,36 @@ function set_interpolated_values_kernel!(
     vert_interpolation_weights,
     vert_bounding_indices,
     field_values,
-)
-    # TODO: Check the memory access pattern. This was not optimized and likely inefficient!
+    ::Val{Nq},
+) where {Nq}
     num_horiz = length(local_horiz_indices)
     num_vert = length(vert_bounding_indices)
     num_fields = length(field_values)
 
-    hindex = (blockIdx().x - Int32(1)) * blockDim().x + threadIdx().x
-    vindex = (blockIdx().y - Int32(1)) * blockDim().y + threadIdx().y
-    findex = (blockIdx().z - Int32(1)) * blockDim().z + threadIdx().z
+    @inbounds begin
+        i_thread =
+            (CUDA.blockIdx().x - Int32(1)) * CUDA.blockDim().x +
+            CUDA.threadIdx().x
+        inds = (num_vert, num_horiz, num_fields)
+
+        1 ≤ i_thread ≤ prod(inds) || return nothing
 
-    totalThreadsX = gridDim().x * blockDim().x
-    totalThreadsY = gridDim().y * blockDim().y
-    totalThreadsZ = gridDim().z * blockDim().z
+        (j, i, k) = CartesianIndices(map(x -> Base.OneTo(x), inds))[i_thread].I
```

Comment:
Shouldn't the order be `(k, j, i)`?

Reply:
Similar with the other ones

Reply:
I think this is correct. The indices should match up with the ranges given to CartesianIndices

Reply:
I think I am still confused by the order: Is the first or the last the fastest index?

Reply:
The first index, returned by `CartesianIndices`, changes the fastest.

permuting the order will change the memory access pattern (which can impact vectorization on the cpu and coalescence on the gpu), but it will not change the correctness.

Reply:
We could follow up with a benchmark if you’d like. For now, I matched the order with fast memory access to `out`

Reply:
Oh, wait, I didn’t. Not sure why I did that order, we can try different orders, but it should probably be performance-driven

Reply:
I probably did that order with the VIJFH datalayouts in mind

Reply:
Okay, maybe we can leave the comment about double checking memory access patterns because this might not be the most effective permutation of indices.